### ISSUE TITLE
Unable to use auto-encrypt feature in secondary consul cluster

### ISSUE THREAD

**Issue by ashwinkupatkar** [hashicorp/consul#10144]
Hello,

I have a federated consul cluster, primary cluster on VM and secondary cluster on k8s.

I have installed my custom CA and verified that the services on primary cluster are able to communicate using the envoy proxy upstream and downstream configurations.

I now use the same consul CA cert and key in k8s cluster (by creating a kubernetes secret).

I have following features enabled:
```

global:
  datacenter: dc-k8s
  image: "hashicorp/consul:1.9.3"
  imageK8S: "hashicorp/consul-k8s:0.24.0"
  enablePodSecurityPolicies: true
  name: "nse"
  tls:
    enabled: true
    verify: true
    enableAutoEncrypt: true
    httpsOnly: true
    caCert:
      secretName: consul-ca-cert
      secretKey: tls.crt
    caKey:
      secretName: consul-ca-key
      secretKey: tls.key

  federation:
    enabled: true

  gossipEncryption:
     secretName: consul-encrypt-key
     secretKey: gossipEncryptionKey

server:
  enabled: true
  storageClass: local-storage          
  extraConfig: |
    {"primary_datacenter": "dc-vm", "primary_gateways":["<primary_mesh_gateway_ip>:8443"]}

connectInject:
  enabled: true
  default: true

controller:
  enabled: true

meshGateway:
  enabled: true
  replicas: 1
  service:
    enabled: true
    type: NodePort
    nodePort: 30002
  consulServiceName: "mesh-gateway"
```

Despite maintaining above configs, I see below error in my consul client logs

`[ERROR] agent.auto_config: AutoEncrypt.Sign RPC failed: addr=10.244.81.135:8300 error="rpcinsecure error making call: rpcinsecure error making call: CA is uninitialized and unable to sign certificates yet: no root certificate"
`

Why does it occur?

Consul: 1.9.3 in both DC
Consul-helm: 0.30.0
envoy: 1.16.0

As per recommendation on [discuss-forum](https://discuss.hashicorp.com/t/unable-to-use-auto-encrypt-feature-in-secondary-consul-cluster/23660) upgraded consul helm chart to 0.31.1 but still the issue persists.

Please find attached kubectl get events output :

[consul-events (2).txt](https://github.com/hashicorp/consul/files/6393483/consul-events.2.txt)

The client pod gets the below output:
```

2021-04-27T20:31:08.911Z [ERROR] agent.auto_config: AutoEncrypt.Sign RPC failed: addr=10.244.81.162:8300 error="rpcinsecure error making call: rpcinsecure error making call: CA is uninitialized and unable to sign certificates yet: no root certificate"
2021-04-27T20:31:08.911Z [ERROR] agent.auto_config: No servers successfully responded to the auto-encrypt request
```

The secondary consul cluster comes up perfectly fine, if i disable auto-encrypt feature… but then it cannot federate with the primary cluster where auto_encrypt is enabled.

What can be the issue ?

**Comment by ashwinkupatkar** [comment]
@ChipV223 can you help me debugging above issue ?

**Comment by dnephin** [comment]
Thank you for reporting this issue! 

If the client agents start before the servers have established a leader then it is expected that the logs could contain a few lines of "AutoEncrypt.Sign RPC failed". That operation will be retried for up to 10 minutes. Once the servers have established a leader, the CA should be initialized, and the errors in the log should stop.

When you were seeing these errors did they stop after a little while, or did you see them continue for 10 minutes? Was a leader elected by the servers? If the errors keep happening after a leader is elected then that does sound like a bug.

**Comment by ashwinkupatkar** [comment]
Hi @dnephin,

Thank you for responding on the issue. A leader do get elected after a certain time and comes up just fine. But the error in the consul client do not go ever.

**Comment by dnephin** [comment]
Thank you for confirming! In the logs for the leader in the secondary DC, do you see anything that says "Failed to initialize Connect CA" or "Successfully initialized the Connect CA" ? You can also check with the CLI command `consul connect ca get-config` to see if the CA manager has been initialized.

It seems like maybe the problem is that the CA provider is not being initialized properly.

**Comment by ashwinkupatkar** [comment]
I do not remember that I have seen info regarding the Ca initialization but what I have seen is consul server certificate do get generated  and they are present in all consul server pods. (checked by landing into server pods). 

I have enabled httpsonly and verify flag so that will mean they have to communicate on https ports and so the certificates would be needed for the communication between servers to take place.

This issue occurs only when auo-encrypt is enabled. Otherwise deployments run smooth as butter.

**Comment by dnephin** [comment]
I'm not as familiar with the helm config, so I had a look at https://github.com/hashicorp/consul-helm/blob/master/templates/server-statefulset.yaml to see how that config is translated into Consul config. From my reading, the Consul servers use the CA certs you specified directly. The servers don't require the CA provider to be initialized. Only the client agents (via auto-encrypt) and connect service-mesh services will make use of the CA provider.

So I think it is still possible that the CA provider is failing to initialize, which would cause this problem. If you are able to reproduce, those two log messages that I referenced above should confirm or disprove that theory. Either way the logs from the leader should contain more information that will help debug this problem. If you are able to share those logs it would help guide us in the right direction.

**Comment by ashwinkupatkar** [comment]
Sure @dnephin that makes sense. I will provide you the logs soon. Currently the setup is being worked upon further by my peer.

As that finishes, I will setup the cluster again using auto encrypt method and provide you the logs.

Thank you!!

**Comment by ashwinkupatkar** [comment]
@dnephin I am soo surprised to see that the issue resolved by itself. I did not do any changes in the config. I just enabled auto-encryp on both primary and secondary clsuter so that I can reproduce the issue today and provide you the relevant logs.... but the setup just worked like a charm. Baffled.

I will close the issue for now. If I face the issue again, will communicate on this thread itself. Thank you for enquiring on the issue.
